macros/match_token.rs

```diff
@@ -0,0 +1,505 @@
+matching, by enforcing the following restrictions on its input:
+
+  - The last pattern must be a variable or the wildcard "_".  In other words
+    it must match everything.
+
+  - Otherwise, ordinary Rust patterns and specific-tag patterns cannot appear
+    after wildcard tag patterns.
+
+  - No tag name may appear more than once.
+
+  - A wildcard tag pattern may not occur in the same arm as any other tag.
+    "<_> <html> => ..." and "<_> </_> => ..." are both forbidden.
+
+  - The right-hand side "else" may only appear with specific-tag patterns.
+    It means that these specific tags should be handled by the last,
+    catch-all case arm, rather than by any wildcard tag arm.  This situation
+    is common in the HTML5 syntax.
+*/
+
+use std::collections::{HashMap, HashSet};
+use std::fmt::Write as FmtWrite;
+use std::fs::File;
+use std::io::{Read, Write};
+use std::path::Path;
+
+#[derive(Clone)]
+struct Source<'a> {
+    src: &'a str,
+}
+
+impl<'a> Source<'a> {
+    fn consume(&mut self, n: usize) -> &'a str {
+        let (before, after) = self.src.split_at(n);
+        self.src = after;
+        before
+    }
+
+    fn find(&mut self, s: &str) -> Option<&'a str> {
+        self.src.find(s).map(|position| {
+            let before = self.consume(position);
+            self.consume(s.len());
+            before
+        })
+    }
+
+    fn consume_if_present(&mut self, s: &str) -> bool {
+        let present = self.src.starts_with(s);
+        if present {
+            self.consume(s.len());
+        }
+        present
+    }
+
+    fn expect(&mut self, s: &str) {
+        assert!(self.consume_if_present(s), "{:?}… does not start with {:?}", &self.src[..50], s);
+    }
+
+    /// Not exactly Rust whitespace, but close enough
+    fn consume_whitespace(&mut self) {
+        while self.src.starts_with(&[' ', '\t', '\n', '\r'][..]) {
+            self.consume(1);
+        }
+    }
+
+    /// Not exactly the syntax of a Rust identifier, but close enough
+    fn consume_ident(&mut self) -> Option<&'a str> {
+        let end = self.src.find(|c: char| !c.is_alphanumeric() && c != '_').unwrap_or(self.src.len());
+        if end > 0 {
+            Some(self.consume(end))
+        } else {
+            None
+        }
+    }
+
+    fn find_top_level(&mut self, start_at: usize, delimeter: u8) -> usize {
+        let mut i = start_at;
+        let bytes = self.src.as_bytes();
+        loop {
+            let b = *bytes.get(i).expect("unbalanced brackets");
+            i += 1;
+            if b == delimeter {
+                return i
+            }
+            match b {
+                b'{' => i = self.find_top_level(i, b'}'),
+                b'[' => i = self.find_top_level(i, b']'),
+                b'(' => i = self.find_top_level(i, b')'),
+                _ => {}
+            }
+        }
+    }
+}
+
+pub fn expand_match_tokens(from: &Path, to: &Path) {
+//    use std::fmt::Write;
```

Comment:
They were temporary, for debugging. I forgot to remove them before committing. Removed now.